exchanges/gateio/gateio.go

```diff
@@ -420,170 +1025,2884 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+		params.Set("offset", strconv.FormatUint(offset, 10))
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	if !from.IsZero() {
+		params.Set("from", strconv.FormatInt(from.Unix(), 10))
+	}
+	if !to.IsZero() && ((!from.IsZero() && to.After(from)) || to.Before(time.Now())) {
+		params.Set("to", strconv.FormatInt(to.Unix(), 10))
+	}
+	var accountBook []AccountBook
+	return accountBook, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, perpetualSwapPrivateEPL, http.MethodGet, optionsAccountbook, params, nil, &accountBook)
+}
+
+// GetUsersPositionSpecifiedUnderlying lists user's positions of specified underlying
+func (g *Gateio) GetUsersPositionSpecifiedUnderlying(ctx context.Context, underlying string) ([]UsersPositionForUnderlying, error) {
+	params := url.Values{}
+	if underlying != "" {
+		params.Set("underlying", underlying)
+	}
+	var response []UsersPositionForUnderlying
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, perpetualSwapPrivateEPL, http.MethodGet, optionsPosition, params, nil, &response)
+}
+
+// GetSpecifiedContractPosition retrieves specified contract position
+func (g *Gateio) GetSpecifiedContractPosition(ctx context.Context, contract currency.Pair) (*UsersPositionForUnderlying, error) {
+	if contract.IsInvalid() {
+		return nil, errInvalidOrMissingContractParam
 	}
-	if !resp.Result {
-		return nil, fmt.Errorf("code:%d message:%s", resp.Code, resp.Message)
+	var response *UsersPositionForUnderlying
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, perpetualSwapPrivateEPL, http.MethodGet,
+		optionsPosition+"/"+contract.String(), nil, nil, &response)
+}
+
+// GetUsersLiquidationHistoryForSpecifiedUnderlying retrieves user's liquidation history of specified underlying
+func (g *Gateio) GetUsersLiquidationHistoryForSpecifiedUnderlying(ctx context.Context, underlying string, contract currency.Pair) ([]ContractClosePosition, error) {
+	if underlying == "" {
+		return nil, errInvalidUnderlying
+	}
+	params := url.Values{}
+	params.Set("underlying", underlying)
+	if contract.IsPopulated() {
+		params.Set("contract", contract.String())
+	}
+	var response []ContractClosePosition
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, perpetualSwapPrivateEPL, http.MethodGet, optionsPositionClose, params, nil, &response)
+}
+
+// PlaceOptionOrder creates an options order
+func (g *Gateio) PlaceOptionOrder(ctx context.Context, arg OptionOrderParam) (*OptionOrderResponse, error) {
+	if arg.Contract == "" {
+		return nil, errInvalidOrMissingContractParam
+	}
+	if arg.OrderSize == 0 {
+		return nil, errInvalidOrderSize
+	}
+	if arg.Iceberg < 0 {
+		arg.Iceberg = 0
+	}
+	if arg.TimeInForce != gtcTIF && arg.TimeInForce != iocTIF && arg.TimeInForce != pocTIF {
+		arg.TimeInForce = ""
+	}
+	if arg.TimeInForce == iocTIF || arg.Price < 0 {
+		arg.Price = 0
+	}
+	arg.Text = "t-" + strconv.FormatInt(g.Websocket.Conn.GenerateMessageID(true), 10)
```

Comment:
This can cause a panic. If the websocket isn't setup and running, these things can be nil:
```
goroutine 51 [running]:
github.com/thrasher-corp/gocryptotrader/exchanges/gateio.(*Gateio).PlaceOptionOrder(0x14000418000, {0x1058c2238, 0x14000038080}, {0x41d3de4356400000, 0x0, {0x140004f4060, 0x18}, {0x0, 0x0}, {0x0, ...}, ...})
	/Users/scottg/go/src/github.com/thrasher-corporation/gocryptotrader/exchanges/gateio/gateio.go:3555 +0x1c8
github.com/thrasher-corp/gocryptotrader/exchanges/gateio.(*Gateio).SubmitOrder(0x14000418000, {0x1058c2238, 0x14000038080}, 0x140001b6370)
	/Users/scottg/go/src/github.com/thrasher-corporation/gocryptotrader/exchanges/gateio/gateio_wrapper.go:1100 +0x3b4
main.testWrappers({0x1058d1780, 0x14000418000}, 0x14000418000, 0x14000092ea0)
	/Users/scottg/go/src/github.com/thrasher-corporation/gocryptotrader/cmd/exchange_wrapper_issues/main.go:606 +0x3d4c
main.main.func1(0x0)
	/Users/scottg/go/src/github.com/thrasher-corporation/gocryptotrader/cmd/exchange_wrapper_issues/main.go:122 +0x104
created by main.main
	/Users/scottg/go/src/github.com/thrasher-corporation/gocryptotrader/cmd/exchange_wrapper_issues/main.go:115 +0x870
```